quinn/src/recv_stream.rs

```diff
@@ -21,8 +21,49 @@ use crate::{
 /// - A variant of [`ReadError`] has been yielded by a read call
 /// - [`stop()`] was called explicitly
 ///
+/// # Closing a stream
+///
+/// When a stream is expected to be closed gracefully the sender should call
+/// [`SendStream::finish`].  However there is no guarantee the connected [`RecvStream`] will
+/// receive the "finished" notification in the same QUIC frame as the last frame which
+/// carried data.
+///
+/// Even if the application layer logic already knows it read all the data because it does
+/// its own framing, it must still read from the [`RecvStream`] until `AsyncRead` signals
+/// End Of File (EOF).  Otherwise it risks inadvertently calling [`RecvStream::stop`] if it
+/// drops the stream.  And calling [`RecvStream::stop`] could result in the connected
+/// [`SendStream::finish`] call failing with a [`WriteError::Stopped`] error.
+///
+/// For example if exactly 10 bytes are to be read, you still need to explicitly read EOF:
+///
+/// ```no_run
+/// # use quinn::{SendStream, RecvStream};
+/// # async fn func(
+/// #     send_stream: &mut SendStream,
+/// #     recv_stream: &mut RecvStream,
+/// # ) -> anyhow::Result<()>
+/// # {
+/// // In the sending task
+/// send_stream.write(&b"0123456789"[..]).await?;
+/// send_stream.finish().await?;
+///
+/// // In the receiving task
+/// let mut buf = [0u8; 10];
+/// let data = recv_stream.read_exact(&mut buf).await?;
+/// if recv_stream.read_to_end(0).await.is_err() {
+///     recv_stream.stop(0u8.into()).ok();
+/// }
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Note that in this example the receiver sends an error to the sender if it received some
+/// unexpected data, forcing the stream to finish in an error state.
```

Comment:
> forcing the stream to finish in an error state.

This might be a bit misleading. If the sender has finished the stream, our stopping it no longer has any meaning, so our having done so is unlikely to be visible. The core function of `stop` is to interrupt any future attempts to transmit on the stream, which is at that point moot.

A more correct statement might be "..., interrupting any further attempts to send data."